### DISCUSSION TITLE
Autenticação e Middlewares

### DISCUSSION THREAD

**Discussion by wellpinho** [frontendbr/forum#1945] [Perguntas e Respostas] [answered]
_Boa tarde devs.

Sei que aqui é frontend mas quem nunca se envolveu com back jogue a sua pedra.

Estou estudando o backend com NodeJS.

Preciso melhorar minhas habilidades com **autenticação de usuários** e com **middlewares**.

Alguma dicas sobre conteúdos relacionados ao tema? Grato pelo seu tempo.__

**"O ignorante critica, o sábio tenta entender."**

**Comment by portothree** [comment, marked as answer]
Boa tarde @wellpinho.

Há inúmeras formas de se lidar com autenticação, mas podemos resumir em duas estratégias:
- Sessão
   Quando o utilizador se autentica, o servidor verifica as credenciais, cria uma sessão (na memória do servidor) e retorna ao utilizador o ID desta sessão. Nos *endpoints* que pedem uma autenticação, o servidor ira verificar pelo ID da sessão e buscar na sua memória qual o utilizador o ID recebido se refere. [Mais sobre sessão com NodeJS](https://petsistemas.ufms.br/index.php/blog/2015/03/25/node-js-autenticacao-com-passport/)
- Tokens
   A diferença com *tokens* é que servidor não guarda nada na sua memória, em vez disso retorna um *token* único, baseado num algoritmo e assinatura da sua escolha e que o servidor também vai esperar receber nos próximos pedidos e verifica o utilizador atual por descodificar o *token* recebido e analisar o *payload*. Alguns exemplos são Bearer Token (Oauth2) e JWT. [Mais sobre tokens com NodeJS](https://imasters.com.br/back-end/autenticacao-json-web-token-jwt-em-node-js)


E com isso pode implementar um *middleware* [express](https://expressjs.com/) ou [connect](https://github.com/senchalabs/connect) e não precisar verificar em cada *controller* a autenticação e autorização do utilizador seja sessão ou *token*.
